### DISCUSSION TITLE
my prisma kept disconnecting

### DISCUSSION THREAD

**Discussion by Baghram** [prisma/prisma#4587] [Q&A] [answered]
Hi, i want to ask some questions, i kept getting error like this when i deployed to a project using openshift,the scenario is like this

- i use postgresql and prisma with next.js
- after a few while after deployed prisma refused to connect with this error

```
"\nInvalid `prisma.user.upsert()` invocation:\n\n\n  Can't reach database server at `postgresql`:`5432`\n\nPlease make sure your database server is running at `postgresql`:`5432`."
```

- i already put the connection_limit=1 at the database_url env for connecting postgres
-- i put the prisma client as a helper like this:

```ts
import { PrismaClient } from '@prisma/client';

const prisma = new PrismaClient();
export default prisma;
```

and my code is mostly standarized like this(i put the await disconnect in the finally at the end of a try catch as per some guide)

```ts
handler.put(async(req,res) => {
  try {
   await prisma.user.findOne(where...)
  }
  catch{}
  finally{
    await prisma.$disconnect()
  }
}
```

the problems is that prisma refused to connect like above,but the postgresql is fine and can be accessed with port-forwarding
any ideas?

**Comment by ryands17** [comment]
@Baghram 👋 
To clarify, does it mean that it works fine for some time and then after a while it cannot connect to the database?

If so, I would suggest removing the `prisma.$disconnect()` call and then checking if this error still occurs.

**Reply by Baghram** [reply]
basically the gist is like that, im currently trying putting the connection_limit to 1 first and if it's still disconnecting i will try to the suggested ones, any ideas why that happens exactly?

**Reply by ryands17** [reply]
@Baghram Sorry, I have a hard time following. Can you try to rephrase what exactly the problem is? Can Prisma connect to the database the first time and at what part are you facing this issue where Prisma cannot connect?

**Reply by Baghram** [reply]
my code is in this kind of format

```ts
handler.get(async (req: any, res: any) => {
  const { token } = req.headers;
  const verifyData = await checktoken(token);
  if (!verifyData) throw new Error('Unauthorized');
  prisma.user
    .findUnique({
      where: {
        id: verifyData.id,
      },
      include: {
        Profile: true,
      },
    })
    .then((result: any) => {
      const data = result;
      delete data?.passwords;
      res.status(200).json({
        message: 'get data success',
        data,
      });
    })
    .catch((err: any) => {
      res.status(400).json({
        message: 'get data failed',
        err,
      });
    })
    .finally(async () => {
      await prisma.$disconnect();
    });
});
```

the problems is that when i deployed  after a while it disconnect with the error like above(1st post), it suddenly died, but when i accidentaly doing an error that push around hundreds of times querying to the server it didn't disconnect, but after a while it disconnects @ryands17

**Reply by Baghram** [reply]
right now im doing the ones that you mention above, where i hide all the await prisma disconnect with the setting of connection_limit=1 in the postgresql env
waiting whether it's gonna be disconnected or not

**Reply by Baghram** [reply]
> @Baghram Sorry, I have a hard time following. Can you try to rephrase what exactly the problem is? Can Prisma connect to the database the first time and at what part are you facing this issue where Prisma cannot connect?

yes, prisma can connect the database the first time, but after a while of usage it disconnects with these kind of errors
Invalid `prisma.user.upsert()` invocation:↵↵↵  Can't reach database server at `postgresql`:`5432`↵↵Please make sure your database server is running at `postgresql`:`5432`.

**Reply by ryands17** [reply]
@Baghram In this case I would suggest removing the `prisma.$disconnect()` call and the checking again if the issue persists. Let me know if it works.

**Reply by Baghram** [reply]
@ryands17 so far there's one concern, if i did not disconnect the prisma, when i run it in local the db wont connect because "there's too many connection already", probably same when i deployed it, it's when my next.js suddenly flicker and shooting hundreds of api

**Reply by ryands17** [reply]
@Baghram A user was facing the same issue. Apparently this solution worked for them:
https://github.com/prisma/prisma/issues/1983#issuecomment-620621213

Let me know if this works for you :)

**Reply by Baghram** [reply]
@ryands17 hi right now im trying to use the one in the 1983(the one that set global) and without the await.prisma.$disconnect(),
but it's showing this in logs

MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 field listeners added to [IncomingForm]. Use emitter.setMaxListeners() to increase limit

either it's something to do with formidable or it's in the prisma, since it's appearing only on recent deploy i think, imma check both of em
any idea?thanks..also if this is disconnect or not imma keep an eye for it for a few days, but in local when i accidentaly hit hundreds of api it didn't refused it dunno why it only appears after i deployed it

**Reply by ryands17** [reply]
@Baghram 👋 
Let me check and get back to you on this :)

**Reply by ryands17** [reply]
I can't seem to reproduce the issue in the below code. Is your setup similar as mine below?

[nextjs-blog.zip](https://github.com/prisma/prisma/files/5724300/nextjs-blog.zip)

**Reply by Baghram** [reply, marked as answer]
my latest deploy it is kinda similar, im using the global prisma in prisma new client from that #1983 and im currently watching it, the incoming form problem is mostly from formidable, but the disconnect thing im kinda watching it right now whether it's gonna disconnect or not since when i mistakenly doing a recursive hell in accessing prisma with axios it doesn't refuse connection but the trouble is avaiable when deploy, that's why im currently on
-using global prisma from that issue,
-removing all await prisma.$disconnect()
-removing all await prisma.$connect()

still watching whether it's gonna disconnect or not and i'll keep updated for it

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
